### ISSUE TITLE
FTBFS on kfreebsd architecture

### ISSUE THREAD

**Issue by bigon** [haproxy/haproxy#1555]
### Detailed Description of the Problem

haproxy fails to build on kfreebsd architecture (a debian architecture with freebsd kernel and the GNU libc)

```
cc -Iinclude  -O2 -g -O2 -ffile-prefix-map=/<<PKGBUILDDIR>>=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type   -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference        -DUSE_KQUEUE    -DUSE_PCRE2 -DUSE_PCRE2_JIT -DUSE_POLL  -DUSE_THREAD     -DUSE_TPROXY   -DUSE_LIBCRYPT  -DUSE_GETADDRINFO -DUSE_OPENSSL -DUSE_LUA  -DUSE_ACCEPT4 -DUSE_CLOSEFROM  -DUSE_SLZ -DUSE_CPU_AFFINITY           -DUSE_PROCCTL   -DUSE_OT  -DUSE_PROMEX   -I/usr/include/lua5.3 -I/usr/include/lua5.3 -DUSE_PCRE2 -DPCRE2_CODE_UNIT_WIDTH=8  -I/usr/include  -Iaddons/ot/include   -DCONFIG_HAPROXY_VERSION=\"2.4.13-1\" -DCONFIG_HAPROXY_DATE=\"2022/02/17\" -c -o src/cpuset.o src/cpuset.c
In file included from include/haproxy/cpuset.h:4,
                 from src/cpuset.c:5:
include/haproxy/cpuset-t.h:40:3: error: #error "No cpuset support implemented on this platform"
   40 | # error "No cpuset support implemented on this platform"
      |   ^~~~~
include/haproxy/cpuset-t.h:45:9: error: unknown type name ‘CPUSET_REPR’
   45 |         CPUSET_REPR cpuset;
      |         ^~~~~~~~~~~
src/cpuset.c: In function ‘ha_cpuset_set’:
src/cpuset.c:33:1: warning: control reaches end of non-void function [-Wreturn-type]
   33 | }
      | ^
src/cpuset.c: In function ‘ha_cpuset_clr’:
src/cpuset.c:48:1: warning: control reaches end of non-void function [-Wreturn-type]
   48 | }
      | ^
src/cpuset.c: In function ‘ha_cpuset_count’:
src/cpuset.c:71:1: warning: control reaches end of non-void function [-Wreturn-type]
   71 | }
      | ^
src/cpuset.c: In function ‘ha_cpuset_ffs’:
src/cpuset.c:95:1: warning: control reaches end of non-void function [-Wreturn-type]
   95 | }
      | ^
src/cpuset.c: In function ‘ha_cpuset_size’:
src/cpuset.c:120:1: warning: control reaches end of non-void function [-Wreturn-type]
  120 | }
      | ^
```

Looking a bit further, it seems that at least `long execfn = getauxval(AT_EXECFN);` can also be a problem, `getauxval` seems to be Linux specific

https://buildd.debian.org/status/package.php?p=haproxy&suite=sid

### Expected Behavior

The software builds

### Steps to Reproduce the Behavior

1. Try to build the software

### Do you have any idea what may have caused this?

The problem seems that the code uses the `__FreeBSD__` constant while the `__FreeBSD_kernel__` should be used instead

### What is your configuration?

```haproxy
Debian kfreebsd port
```

### Output of `haproxy -vv`

```plain
N/A
```

**Comment by wtarreau** [comment]
Thanks! Just to be sure, which exact version did you try to build ? There were recent changes on this code in 2.6-dev and I suspect that's related to them one way or another but would like to know if the issue is with new or older versions.

**Comment by TimWolla** [comment]
> Thanks! Just to be sure, which exact version did you try to build ?

You can find the full log here: https://buildd.debian.org/status/fetch.php?pkg=haproxy&arch=kfreebsd-amd64&ver=2.4.13-1&stamp=1645091818&raw=0. It's 2.4.13.

**Comment by bigon** [comment]
I tried locally and 2.5.2 also seems to fail the same way

**Comment by wtarreau** [comment]
OK thanks to you both. @bigon, a few more questions, given that these were changed in 2.6-dev, could you please give this a try ? The rationale behind this is that David Carlier told us that FreeBSD recently changed their API (in v14) to match linux's, and he proposed a different approach. Maybe we'll need to backport that. In addition, I've never tried debian/kfreebsd, but the FreeBSD target has a strong assumption that the libc also is the FreeBSD one. We can possibly change this if more flexibility is needed, but then we need to know what has to be detected and how to choose what API to use.

Also, did you ever manage to build a version on this platform, and if so, which one ? And if you add `USE_CPU_AFFINITY=` to your make command line, does it build ?

**Comment by bigon** [comment]
kfreebsd port is still using freebsd kernel 10 or 11

> the FreeBSD target has a strong assumption that the libc also is the FreeBSD one

Yes indeed, most of the problems are coming from that

With the proposed following patches, things are going further, but the build fails later:


```diff
Index: haproxy-2.5.2/include/haproxy/cpuset-t.h
===================================================================
--- haproxy-2.5.2.orig/include/haproxy/cpuset-t.h
+++ haproxy-2.5.2/include/haproxy/cpuset-t.h
@@ -6,12 +6,13 @@
 
 #if defined(__FreeBSD__) || defined(__DragonFly__) || defined(__NetBSD__)
 #include <sys/param.h>
-#ifdef __FreeBSD__
+#endif
+
+#if defined(__FreeBSD__) || defined(__FreeBSD_kernel__)
 #include <sys/_cpuset.h>
 #include <sys/cpuset.h>
 #include <strings.h>
 #endif
-#endif
 
 #include <haproxy/api-t.h>
 
@@ -20,7 +21,7 @@
 # define CPUSET_REPR cpu_set_t
 # define CPUSET_USE_CPUSET
 
-#elif defined(__FreeBSD__) || defined(__NetBSD__)
+#elif defined(__FreeBSD__) || defined(__NetBSD__) || defined(__FreeBSD_kernel__)
 
 # define CPUSET_REPR cpuset_t
 
Index: haproxy-2.5.2/src/tools.c
===================================================================
--- haproxy-2.5.2.orig/src/tools.c
+++ haproxy-2.5.2/src/tools.c
@@ -43,7 +43,7 @@ extern void *__elf_aux_vector;
 #include <netinet/in.h>
 #include <arpa/inet.h>
 
-#if defined(__GLIBC__) && (__GLIBC__ > 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ >= 16))
+#if defined(__linux__) && (defined(__GLIBC__) && (__GLIBC__ > 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ >= 16)))
 #include <sys/auxv.h>
 #endif
 
@@ -4781,7 +4781,7 @@ const char *get_exec_path()
 {
 	const char *ret = NULL;
 
-#if defined(__GLIBC__) && (__GLIBC__ > 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ >= 16))
+#if defined(__linux__) && (defined(__GLIBC__) && (__GLIBC__ > 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ >= 16)))
 	long execfn = getauxval(AT_EXECFN);
 
 	if (execfn && execfn != ENOENT)
```

**Comment by wtarreau** [comment]
I think we must be careful not to mix everything like this. First, "freebsd" is both a kernel and a full-blown operating system providing its own libc, so we'll have to declare a new build target "freebsd-glibc" to designate the combination you're using, just like we have "linux-glibc" that differs from "linux-musl" for example. Once we have this we can see what build options that currently depend on linux/freebsd are in fact dependent on the libc and ought to depend on glibc (and conversely). That's also why I'm interested in a response to the questions above. I suspect that what David did for freebsd14 needs to be generalized to (glibc||freebsd>=14) as it acts at the libc layer for example (and possibly ought to be backported). But there are possibly other areas that need to be checked as you found for the AUX vector.

**Comment by devnexen** [comment]
Would the diff, inspired from @bigon 's , work ?

```
diff --git a/Makefile b/Makefile
index b97bbf5..5b249bd 100644
--- a/Makefile
+++ b/Makefile
@@ -177,8 +177,8 @@ DOCDIR = $(PREFIX)/doc/haproxy
 #### TARGET system
 # Use TARGET=<target_name> to optimize for a specific target OS among the
 # following list (use the default "generic" if uncertain) :
-#    linux-glibc, linux-glibc-legacy, linux-musl, solaris, freebsd, dragonfly,
-#    openbsd, netbsd, cygwin, haiku, aix51, aix52, aix72-gcc, osx, generic,
+#    linux-glibc, linux-glibc-legacy, linux-musl, solaris, freebsd, freebsd-glibc,
+#    dragonfly, openbsd, netbsd, cygwin, haiku, aix51, aix52, aix72-gcc, osx, generic,
 #    custom
 TARGET =
 
@@ -412,6 +412,13 @@ ifeq ($(TARGET),freebsd)
     USE_ACCEPT4 USE_CLOSEFROM USE_GETADDRINFO USE_PROCCTL)
 endif
 
+# kFreeBSD glibc
+ifeq ($(TARGET),freebsd-glibc)
+  set_target_defaults = $(call default_opts, \
+    USE_POLL USE_TPROXY USE_LIBCRYPT USE_THREAD USE_CPU_AFFINITY USE_KQUEUE   \
+    USE_ACCEPT4 USE_GETADDRINFO USE_CRYPT_H USE_DL)
+endif
+
 # DragonFlyBSD 4.3 and above
 ifeq ($(TARGET),dragonfly)
   set_target_defaults = $(call default_opts, \
diff --git a/include/haproxy/cpuset-t.h b/include/haproxy/cpuset-t.h
index 5f812aa..984df8d 100644
--- a/include/haproxy/cpuset-t.h
+++ b/include/haproxy/cpuset-t.h
@@ -16,7 +16,7 @@
 
 #include <haproxy/api-t.h>
 
-#if defined(__linux__) || defined(__DragonFly__)
+#if defined(__linux__) || defined(__DragonFly__) || defined(__FreeBSD_kernel__)
 
 # define CPUSET_REPR cpu_set_t
 # define CPUSET_USE_CPUSET
diff --git a/src/tools.c b/src/tools.c
index c48a969..f62032b 100644
--- a/src/tools.c
+++ b/src/tools.c
@@ -43,7 +43,7 @@ extern void *__elf_aux_vector;
 #include <netinet/in.h>
 #include <arpa/inet.h>
 
-#if defined(__GLIBC__) && (__GLIBC__ > 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ >= 16))
+#if defined(__linux__) && defined(__GLIBC__) && (__GLIBC__ > 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ >= 16))
 #include <sys/auxv.h>
 #endif
 
@@ -4791,7 +4791,7 @@ const char *get_exec_path()
 {
 	const char *ret = NULL;
 
-#if defined(__GLIBC__) && (__GLIBC__ > 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ >= 16))
+#if defined(__linux__) && defined(__GLIBC__) && (__GLIBC__ > 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ >= 16))
 	long execfn = getauxval(AT_EXECFN);
 
 	if (execfn && execfn != ENOENT)
```

**Comment by wtarreau** [comment]
Thanks David. If it works, I'm fine with merging this one, it matches exactly what I'd have in mind.

**Comment by wtarreau** [comment]
@bigon did you have the opportunity/time to test @devnexen 's patch above ? The earlier we merge this, the better. Thanks.

**Comment by bigon** [comment]
Hello,

I tired the patch and now I've the follwoing failure:

```
src/auth.c:43:28: error: variable 'crypt_data' has initializer but incomplete type
   43 | static THREAD_LOCAL struct crypt_data crypt_data = { .initialized = 0 };
      |                            ^~~~~~~~~~
src/auth.c:43:55: error: 'struct crypt_data' has no member named 'initialized'
   43 | static THREAD_LOCAL struct crypt_data crypt_data = { .initialized = 0 };
      |                                                       ^~~~~~~~~~~
src/auth.c:43:69: warning: excess elements in struct initializer
   43 | static THREAD_LOCAL struct crypt_data crypt_data = { .initialized = 0 };
      |                                                                     ^
src/auth.c:43:69: note: (near initialization for 'crypt_data')
src/auth.c: In function 'check_user':
src/auth.c:260:22: warning: implicit declaration of function 'crypt_r'; did you mean 'crypt'? [-Wimplicit-function-declaration]
  260 |                 ep = crypt_r(pass, u->pass, &crypt_data);
      |                      ^~~~~~~
      |                      crypt
src/auth.c:260:20: warning: assignment to 'const char *' from 'int' makes pointer from integer without a cast [-Wint-conversion]
  260 |                 ep = crypt_r(pass, u->pass, &crypt_data);
      |                    ^
src/auth.c: At top level:
src/auth.c:43:39: error: storage size of 'crypt_data' isn't known
   43 | static THREAD_LOCAL struct crypt_data crypt_data = { .initialized = 0 };
      |                                       ^~~~~~~~~~
make[2]: *** [Makefile:962: src/auth.o] Error 1
make[2]: Leaving directory '/tmp/haproxy-2.4.14'
make[1]: *** [debian/rules:50: override_dh_auto_build-arch] Error 2
make[1]: Leaving directory '/tmp/haproxy-2.4.14'
make: *** [debian/rules:45: build-arch] Error 2
dpkg-buildpackage: error: debian/rules build-arch subprocess returned exit status 2
```

**Comment by wtarreau** [comment]
Maybe this one can be overcome by just appending `USE_LIBCRYPT=` to your "make" command line (in which case we'd just drop it from David's patch). If it doesn't work, then I guess we'll have to keep this cooking for longer for the time it takes to port to that platform.

**Comment by bigon** [comment]
With `USE_LIBCRYPT=` I get the following:

```
cc -Wl,-z,relro -Wl,-z,now -Wl,--as-needed -o haproxy src/slz.o src/ev_poll.o src/ev_kqueue.o src/cpuset.o src/ssl_sample.o src/ssl_sock.o src/ssl_crtlist.o src/ssl_ckch.o src/ssl_utils.o src/cfgparse-ssl.o src/jwt.o src/hlua.o src/hlua_fcn.o addons/promex/service-prometheus.o addons/ot/src/cli.o addons/ot/src/conf.o addons/ot/src/event.o addons/ot/src/filter.o addons/ot/src/group.o addons/ot/src/http.o addons/ot/src/opentracing.o addons/ot/src/parser.o addons/ot/src/pool.o addons/ot/src/scope.o addons/ot/src/util.o src/mux_h2.o src/mux_fcgi.o src/http_ana.o src/mux_h1.o src/tcpcheck.o src/stream.o src/stats.o src/server.o src/flt_spoe.o src/stick_table.o src/tools.o src/sample.o src/log.o src/peers.o src/resolvers.o src/backend.o src/cfgparse.o src/http_htx.o src/cli.o src/proxy.o src/pattern.o src/connection.o src/check.o src/cfgparse-listen.o src/cache.o src/haproxy.o src/http_act.o src/http_fetch.o src/stream_interface.o src/dns.o src/listener.o src/http_client.o src/vars.o src/tcp_rules.o src/debug.o src/sink.o src/server_state.o src/filters.o src/h2.o src/fcgi-app.o src/task.o src/payload.o src/h1_htx.o src/mjson.o src/h1.o src/map.o src/flt_http_comp.o src/cfgparse-global.o src/mux_pt.o src/htx.o src/flt_trace.o src/acl.o src/trace.o src/tcp_sample.o src/mqtt.o src/lb_chash.o src/mworker.o src/activity.o src/sock.o src/session.o src/queue.o src/tcp_act.o src/ring.o src/proto_tcp.o src/fd.o src/http_rules.o src/channel.o src/http.o src/extcheck.o src/compression.o src/thread.o src/frontend.o src/raw_sock.o src/pool.o src/http_conv.o src/uri_normalizer.o src/lb_fwrr.o src/xprt_handshake.o src/sha1.o src/mailers.o src/ebmbtree.o src/applet.o src/proto_sockpair.o src/lb_fwlc.o src/errors.o src/lb_fas.o src/action.o src/proto_uxst.o src/cfgcond.o src/protocol.o src/proto_udp.o src/lb_map.o src/hpack-dec.o src/fix.o src/ev_select.o src/arg.o src/sock_unix.o src/sock_inet.o src/signal.o src/proto_uxdg.o src/mworker-prog.o src/fcgi.o src/eb32sctree.o src/clock.o src/chunk.o src/cfgparse-tcp.o src/cfgdiag.o src/shctx.o src/lru.o src/eb64tree.o src/eb32tree.o src/cfgparse-unix.o src/regex.o src/hpack-tbl.o src/ebimtree.o src/base64.o src/uri_auth.o src/time.o src/ebsttree.o src/ebistree.o src/dynbuf.o src/auth.o src/wdt.o src/pipe.o src/http_acl.o src/hpack-huff.o src/hpack-enc.o src/dict.o src/init.o src/freq_ctr.o src/ebtree.o src/hash.o src/dgram.o src/version.o   -lpthread  -lssl -lcrypto -Wl,--export-dynamic  -llua5.3 -lm -L/usr/lib -L/usr/lib -lpcre2-8 -lpcre2-posix -lopentracing-c-wrapper 
/usr/bin/ld: src/tools.o: undefined reference to symbol 'dlsym@@GLIBC_2.3'
/usr/bin/ld: /lib/x86_64-kfreebsd-gnu/libdl.so.2: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:959: haproxy] Error 1
make[2]: Leaving directory '/tmp/haproxy-2.5.4'
make[1]: *** [debian/rules:51: override_dh_auto_build-arch] Error 2
make[1]: Leaving directory '/tmp/haproxy-2.5.4'
make: *** [debian/rules:46: build-arch] Error 2
dpkg-buildpackage: error: debian/rules build-arch subprocess returned exit status 2
```

**Comment by devnexen** [comment]
A bit perplex by what you encounter, if that makes a difference I downloaded kfreebsd from [here](https://wiki.debian.org/Debian_GNU/kFreeBSD).

**Comment by wtarreau** [comment]
Strange. You have USE_DL which is set in the patch, yet no -ldl on the linker's command line, while it's systematic in the makefile. What is the exact "make" command line you're passing to it ?

**Comment by bigon** [comment]
I think you should explicitly link against `libdl.so`

**Comment by bigon** [comment]
```
make[1]: Entering directory '/tmp/haproxy-2.5.4'
make V=1 DESTDIR=debian/haproxy PREFIX=/usr IGNOREGIT=true MANDIR=/usr/share/man DOCDIR=/usr/share/doc/haproxy USE_PCRE2=1 USE_PCRE2_JIT=1 USE_OPENSSL=1 USE_SLZ=1 USE_LUA=1 USE_LIBCRYPT= USE_PROMEX=1 USE_OT=1 LUA_INC=/usr/include/lua5.3 EXTRA=admin/halog/halog TARGET=freebsd DEBUG_CFLAGS="-g -O2 -ffile-prefix-map=/tmp/haproxy-2.5.4=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2" LDFLAGS="-Wl,-z,relro -Wl,-z,now -Wl,--as-needed" VERSION="2.5.4" SUBVERS="-1" VERDATE="2022/02/25"
make[2]: Entering directory '/tmp/haproxy-2.5.4'
```

**Comment by devnexen** [comment]
wrong target I see. `freebsd-glibc` should be passed. Basically also `USE_LIBCRYPT=` should be removed then.

**Comment by bigon** [comment]
Oupsie

So yes, it builds fine with changing the target to `freebsd-glibc` (and without adding `USE_LIBCRYPT=`)

**Comment by wtarreau** [comment]
Cool! David, could you please send me a clean patch with this, referencing this issue for tracking ? Please clearly mention in the commit message the reasons for each of the 3 changes in the include files, as these can be useful when adding new platforms, particularly if we have to refine them without breaking this one!

Thank you guys!

**Comment by wtarreau** [comment]
@bigon this was backported to 2.5 for your convenience. I'm not sure it's really needed to go further.

**Comment by wtarreau** [comment]
Reopening as this broke the build on the real FreeBSD :-(

  https://github.com/haproxy/haproxy/runs/5465299996

It seems that __FreeBSD_kernel__ is defined there as well and that we ought to rely on that+glibc instead. Actually we should probably change the __linux__ test for __GLIBC__ in the cpu_set, though we would probably lose musl support. But that's definitely a libc API.

**Comment by devnexen** [comment]
to be more precise it fails on FreeBSD < 14, I have a FreeBSD 13 VM somewhere should be straightforward.

**Comment by wtarreau** [comment]
I tested here on a VM on 11.3 and it does not deefine this __FreeBSD_kernel__:
```
 clang -dM -E - </dev/null |grep -i freebsd
#define __FreeBSD__ 11
#define __FreeBSD_cc_version 1100512
#define __VERSION__ "4.2.1 Compatible FreeBSD Clang 8.0.0 (tags/RELEASE_800/final 356365)"
```
So it probably appeared later. I thikn that in the worst case freebsd_kernel&&glibc will suffice.

**Comment by wtarreau** [comment]
Ah yes it does, it's inherited from `sys/param.h`:
```
$ clang -dM -E - -include /usr/include/sys/param.h </dev/null |grep -i freebsd
#define __FreeBSD__ 11
#define __FreeBSD_cc_version 1100512
#define __FreeBSD_kernel__ 
#define __FreeBSD_version 1103000
#define __VERSION__ "4.2.1 Compatible FreeBSD Clang 8.0.0 (tags/RELEASE_800/final 356365)"
```
At least there's an explanation now.

**Comment by wtarreau** [comment]
So in short we really need to focus only on `(defined(__FreeBSD__) && defined(__GLIBC__))` at the end of that line.

**Comment by wtarreau** [comment]
Since it's fixed and backported to 2.5, let's close it.

**Comment by bigon** [comment]
Thanks
